odbc/src/util.rs

```diff
@@ -0,0 +1,80 @@
+use crate::handles::{MongoHandle, ODBCError};
+use odbc_sys::{Handle, HandleType, SmallInt, SqlReturn, WChar};
+use std::{cmp::min, ptr::copy};
+
+/// add_diag_info creates a new `ODBCError` object and appends it to the
+/// given handle's `errors` field.
+pub fn add_diag_info(
+    handle_type: HandleType,
+    handle: Handle,
+    sql_state: String,
+    error_message: String,
+    native_err_code: i32,
+    component: String,
+) -> Result<(), ()> {
+    let error = ODBCError {
+        sql_state,
+        error_message,
+        native_err_code,
+        component,
+    };
+    match handle_type {
+        HandleType::Env => {
+            let env = unsafe { (*(handle as *mut MongoHandle)).as_env().ok_or(())? };
+            let mut env_contents = env.write().unwrap();
+            env_contents.errors.push(error);
+        }
+        HandleType::Dbc => {
+            let dbc = unsafe { (*(handle as *mut MongoHandle)).as_connection().ok_or(())? };
+            let mut dbc_contents = dbc.write().unwrap();
+            dbc_contents.errors.push(error);
+        }
+        HandleType::Stmt => {
+            let stmt = unsafe { (*(handle as *mut MongoHandle)).as_statement().ok_or(())? };
+            let mut stmt_contents = stmt.write().unwrap();
+            stmt_contents.errors.push(error);
+        }
+        HandleType::Desc => {
+            let desc = unsafe { (*(handle as *mut MongoHandle)).as_descriptor().ok_or(())? };
+            let mut desc_contents = desc.write().unwrap();
+            desc_contents.errors.push(error);
+        }
+    };
+    Ok(())
+}
+
+/// set_sql_state writes the given sql state to the [`output_ptr`].
+pub fn set_sql_state(mut sql_state: String, output_ptr: *mut WChar) {
```

Comment:
nit: I don't think there's any need for this function to take ownership of `sql_string`, especially because you're just cloning the arg at the call site! Even better, this function should probably accept some sort of `SQLState` enum instead of just any string

Reply:
Done